kaitaistruct.py

```diff
@@ -15,51 +20,196 @@
-    def from_io(cls, io):
-        return cls(KaitaiStream(io))
+    def from_io(cls, io: _IOBase) -> "KaitaiStruct":
+        return cls.from_any(io)
 
 
-class KaitaiStream(object):
-    def __init__(self, io):
-        self._io = io
-        self.align_to_byte()
+class IKaitaiDownStream(ABC):
+    __slots__ =("_io",)
+
+    def __init__(self, _io: typing.Any):
+        self._io = _io
+
+    @abstractmethod
+    @property
+    def is_entered(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def __enter__(self):
+        raise NotImplementedError()
+
+    def __exit__(self, *args, **kwargs):
+        if self.is_entered:
+            self._io.__exit__(*args, **kwargs)
+            self._io = None
+
+
+class KaitaiIODownStream(IKaitaiDownStream):
+    __slots__ = ()
+    def __init__(self, data: typing.Any):
+        super().__init__(data)
+
+    @property
+    def is_entered(self):
+        return isinstance(self._io, _IOBase)
 
     def __enter__(self):
+        if not self.is_entered
+            self._io = open(self._io).__enter__()
+        return self
+
+
+class KaitaiBytesDownStream(KaitaiIODownStream):
+    __slots__ = ()
+    def __init__(self, data: bytes):
+        super().__init__(data)
+
+
+class KaitaiFileSyscallDownStream(KaitaiIODownStream):
+    __slots__ = ()
+    def __init__(self, io: typing.Union[Path, str, _IOBase]):
+        if isinstance(io, str):
+            io = Path(io)
+        super().__init__(io)
+
+
+class KaitaiFileMapDownStream(IKaitaiFileDownStream):
+    __slots__ = ("file",)
+    def __init__(self, io: typing.Union[Path, str, _IOBase]):
+        super().__init__(None)
+        self.file = KaitaiFileSyscallDownStream(io)
+
+    @property
+    def is_entered(self):
+        return isinstance(self._io, mmap.mmap)
+
+    def __enter__(self):
+        self.file = self.file.__enter__()
+        self._io = mmap.mmap(self.file.file.fileno(), 0, access=mmap.ACCESS_READ).__enter__()
         return self
 
     def __exit__(self, *args, **kwargs):
-        self.close()
+        super().__exit__(*args, **kwargs)
+        if self.file is not None:
+            self.file.__exit__(*args, **kwargs)
+            self.file = None
+
+
+def get_file_down_stream(path: Path, *args, use_mmap: bool=True, **kwargs) -> IKaitaiDownStream:
+    if use_mmap:
+        cls = KaitaiFileMapDownStream
+    else:
+        cls = KaitaiFileSyscallDownStream
+    return cls(path, *args, **kwargs)
+
+
+downstreamMapping = {
+    bytes: KaitaiBytesDownStream,
+    BytesIO: KaitaiBytesDownStream,
+    str: get_file_down_stream,
+    Path: get_file_down_stream,
+    _io._BufferedIOBase: get_file_down_stream,
+}
+
+
+def get_downstream_ctor(x) -> typing.Type[IKaitaiDownStream]:
+    ctor = downstreamMapping.get(t, None)
+    if ctor:
+        return ctor
+    types = t.mro()
+    for t1 in types[1:]:
+        ctor = downstreamMapping.get(t1, None)
+        if ctor:
+            downstreamMapping[t] = ctor
+            return ctor
```

Comment:
Why not just `isinstance`?

Reply:
Complexity. Doing `isinstance` would mean we would have to walk all the types and check each, this is O(n) and maybe more depending on inheritance impl. I have tried to optimize it by a lookup in a dict that should be O(1).